doc/debug-companion.md

```diff
@@ -1 +1,171 @@
+
+It depends on an underlying network transport library (their most popular seems to be Asio) to interact with the host operating system, but includes all of the necessary pieces to upgrade an HTTP connection and build/parse the WebSocket frames.
+
+Although it claims to have basic HTTP support, its usage hasn't been investigated for this purpose.  Roughly it looks like it just supports passthrough of HTTP headers for further processing.
+
+#### HTTP
+
+The HTTP endpoint services some basic requests from the frontend:
+* /json or /json/list
+  * Returns a JSON object containing the debuggable endpoints
+* /json/protocol
+  * Returns a JSON object containing the current protocol specification
+* /json/version
+  * Returns a JSON object containing the current "Browser" version and "Protocol-Version"
+* /json/activate/\<id\>
+  * Unsure what the purpose is, Node.js just seems to return "Target activated" if it's already active or nothing otherwise.
+
+These are basic GET requests and shouldn't require any complex parsing (HTTP headers for instance).  It's not clear if any of these should be forwarded to the companion for responses, at the moment it seems like these are all host-specific (except maybe the protocol).
+
+## Code Layout and Packaging
+
+### Debugger Companion
+
+The debugger companion only uses the public (but still experimental) API surface of ChakraCore.  This allows some flexibility in where the code should live:
+* A separate repo which builds against the public interface of ChakraCore as an external dependency.
+* As part of the existing ChakraCore repo where it can build against a given version of ChakraCore.
+
+Since the debugger companion will be tightly coupled with the (public, but experimental) interface of ChakraCore, it makes sense for it to live alongside the ChakraCore version for which it was built.  It's beyond the scope of this project to try and support all older versions but going forward we will still want to innovate in the debugger space and will not want to manage these changes externally.
+The code for the companion will live inside of ChakraCore and build a LIB (Debugger.Companion.lib) and a DLL (Debugger.Companion.dll).
+
+Paths:
+* \<root\>\core\lib\Debugger\Companion\
+  * Debugger.Companion.vcxproj
+  * DebuggerCompanion.h
+  * \<core implementation files\>
+* \<root\>\core\bin\Debugger\Companion\Debugger.Companion.vcxproj
+  * DebuggerCompanion.def
+
+The primary consumption will be via the DLL which will use only JSRT APIs to interact with the engine.
+
+### Platform Implementation
+
+Because of the third-party dependencies for this part of the code, it likely should not live alongside ChakraCore.
+
+Possible locations:
+* Chakra-Samples
+
+### Sample Code
```

Comment:
"Sample" sounds like it isn't really functional.  Perhaps rather than "Sample Code" we should call it the "Standard Debugger" or something like that.

Reply:
The sample in this case is a host app which combines all of the pieces along with a script loader to demonstrate the functionality.  There's no concept of "standard" here since the host is up to the implementer.